### ISSUE TITLE
Discuss whether or not we should adjust hotkey assignments

### ISSUE THREAD

**Issue by xezon** [TheSuperHackers/GeneralsGamePatch#242]
Discuss whether or not we should adjust hotkey assignments. Hotkeys are a mess. Different across languages. No logical key assignments. Missing key assignments. It is difficult and time consuming to tweak hotkeys with the tools we have.

Related task: (#19)

**Comment by xezon** [comment]
Right now I am thinking this is one of the things we should not touch for the following reasons:
* We have no good tools to process all languages properly
* There are multiple hotkey sets available for English language that we could offer as optional Addon in the Patch Loader
* Force changing hotkeys can be a major disruption for all players who use original hotkeys
* Players who want to use custom Hotkeys likely already have them setup locally
* A tool to customize hotkeys on user install would be the best way to address this

**Comment by Kesanov** [comment]
- > It is difficult and time consuming to tweak hotkeys with the tools we have.

    It is possible to generate textual representation of `generals.csf`, which significantly simplifies hotkey and tooltip modification.
    It is also much more friendly to git diffs.
    The generated file is named `generals.str` and you can see it in my PR: https://github.com/xezon/GeneralsGamePatch/pull/243
- > Force changing hotkeys can be a major disruption for all players who use original hotkeys

    I totally agree with this. We shouldn't touch existing hotkeys. They are not great, but the majority is used to them.
    However, not all buttons and commands have hotkey assigned.  
    I think there is no harm in assigning hotkeys to those commands.

Known unassigned hotkeys:
- Sell button in English localization.
- Select all Aircraft in German localization.

**Comment by commy2** [comment]
[1.04 generals.str](https://raw.githubusercontent.com/commy2/zerohour/master/Data/generals.str)

**Comment by commy2** [comment]
All text changes, including hotkey changes *must* be optional and super obvious to opt-out, because otherwise we interfer with custom hotkeys.

**Comment by xezon** [comment]
> All text changes [...] must be optional [...], because otherwise we interfer with custom hotkeys

I think this is not accurate. If we just change text and keep hotkeys and embed this into a meaningful load order, then this should be of no negative consequence.

**Comment by xamorish** [comment]
> Force changing hotkeys can be a major disruption for all players who use original hotkeys
> Players who want to use custom Hotkeys likely already have them setup locally

Custom keybindings should be optional and easy to configure.

> We have no good tools to process all languages properly
> There are multiple hotkey sets available for English language that we could offer as optional Addon in the Patch Loader

If we would've had a keybinding system (in launcher, or whatever), we could just iterate through all required strings in the active `generals.csf` file of the game and append the bound letter at the end of them.

Example:

Before:
`Patriot Ba&ttery`
't' is the bound button to construct a Patriot Battery.

After:
`Patriot Battery (&P)`
Makes 'P' the bound button to construct a Patriot Battery.

This eliminates the need to care about what language the consumer is using, because we only need the end of the string, not its initial content.

`ЗРБ 'Патриот' (&P)`
Still works the same.

We'd need to find a way to read/write valid data to .csf files though.
All of this should be done after backing up the original `generals.csf` file to avoid user losing his original bindings and providing them a way to recover.

**Comment by xezon** [comment]
Regarding tool, I was also dumping some ideas in this topic.
https://www.gamereplays.org/community/index.php?showtopic=1052300

Regarding CSF file format, I looked at it in hex editor and figured it like so:

```
HEADER:
4 bytes: ' FSC': file type identifier
4 bytes: 3: unknown value, maybe file type version
4 bytes: 0x1916: 6422, could be number of string table entries
4 bytes: 0x1915: 6421, similar to number above, not clear what the difference is
8 bytes: all 0: unknown

BODY (repeats):
4 bytes: ' LBL': keyword identifier
4 bytes: 1: unknown value, maybe number of keywords? seems to be 1 always.
4 bytes: length of ascii string (N)
N bytes: ascii string
4 bytes: ' RTS': locale string identifier
4 bytes: length of unicode string (N)
N bytes: unicode string: XOR'ed by 0xFF
```

Thyme also has (better) information on CSF file:
https://github.com/TheAssemblyArmada/Thyme/wiki/Compiled-String-File-Format

**Comment by commy2** [comment]
> We'd need to find a way to read/write valid data to .csf files though.

For what purpose?

**Comment by Kesanov** [comment]
> Regarding tool, I was also dumping some ideas in this topic.

For what it is worth, I had in plan to create a script that automatically assigns hotkeys based on button positions.

That is, the command buttons 
```
1234567
89.....
```
Would get autogenerated hotkeys based on your keyboard layout. For example, QUERTY would get:
```
ASDFGHJ
ZXCVBNM
```
Or perhaps with modified ControlBar that displays buttons in 3 rows:
```
QWERT
ASDFG
ZXCV
```

I believe it to be the optimal hotkey layout for any ZeroHour newbie (who plays the game for first time).

------

Note, this requires some further modifications, like reassigning current `QWEFH` to `SHIFT+QWEFH`

**Comment by xamorish** [comment]
> 
> 
> > We'd need to find a way to read/write valid data to .csf files though.
> 
> For what purpose?

To edit the keybinds? They're stored in the `generals.csf` localization file. The launcher can access that file and write its own data into it if it had a keybinding system.

**Comment by xezon** [comment]
Imo launcher should not contain Hotkey Editor. Hotkey Editor should be separate tool. But we can could have button in Launcher to start Hotkey Editor.

**Comment by xamorish** [comment]
> 
> 
> Imo launcher should not contain Hotkey Editor. Hotkey Editor should be separate tool. But we can could have button in Launcher to start Hotkey Editor.

Yeah, i agree on this. The launcher could pass the Hotkey Editor the path of the `generals.csf` file to edit in the startup parameters, and the Editor would work on its own from there. Provides greater user coverage due to usability outside of the Launcher context.

**Comment by xamorish** [comment]
> 
> 
> Regarding tool, I was also dumping some ideas in this topic.
> https://www.gamereplays.org/community/index.php?showtopic=1052300
> 
> Regarding CSF file format, I looked at it in hex editor and figured it like so:
> 
> ```
> HEADER:
> 4 bytes: ' FSC': file type identifier
> 4 bytes: 3: unknown value, maybe file type version
> 4 bytes: 0x1916: 6422, could be number of string table entries
> 4 bytes: 0x1915: 6421, similar to number above, not clear what the difference is
> 8 bytes: all 0: unknown
> 
> BODY (repeats):
> 4 bytes: ' LBL': keyword identifier
> 4 bytes: 1: unknown value, maybe number of keywords? seems to be 1 always.
> 4 bytes: length of ascii string (N)
> N bytes: ascii string
> 4 bytes: ' RTS': locale string identifier
> 4 bytes: length of unicode string (N)
> N bytes: unicode string: XOR'ed by 0xFF
> ```
> 
> Thyme also has (better) information on CSF file:
> https://github.com/TheAssemblyArmada/Thyme/wiki/Compiled-String-File-Format

Very nice material.

**Comment by xezon** [comment]
> Yeah, i agree on this. The launcher could pass the Hotkey Editor the path of the generals.csf file to edit in the startup parameters, and the Editor would work on its own from there. Provides greater user coverage due to usability outside of the Launcher context.

I don't think that is a good idea. A Hotkey Editor ideally is self contained and allows the user to pick .csf file to edit. Hotkey Editor needs 2 functionalities for this:
a) Be able to browse into .big files
b) Be able to auto discover first generals.csf file (extracted or in .big)

This way flexible functionality can be implemented to allow user find and edit .csf file from within Editor without touching anything else, no matter if it is launched individually or from external Launcher app.

**Comment by xamorish** [comment]
> 
> 
> > Yeah, i agree on this. The launcher could pass the Hotkey Editor the path of the generals.csf file to edit in the startup parameters, and the Editor would work on its own from there. Provides greater user coverage due to usability outside of the Launcher context.
> 
> I don't think that is a good idea. A Hotkey Editor ideally is self contained and allows the user to pick .csf file to edit. Hotkey Editor needs 2 functionalities for this:
> a) Be able to browse into .big files
> b) Be able to auto discover first generals.csf file (extracted or in .big)
> 
> This way flexible functionality can be implemented to allow user find and edit .csf file from within Editor without touching anything else, no matter if it is launched individually or from external Launcher app.

I don't think i made myself clear enough: I didn't mean that the Hotkey Editor shouldn't have an option to select different .csf files, i said that it should be started with the user's active `generals.csf` file being pre-selected. It wouldn't be usable outside of the Launcher's context if the user couldn't select files himself. Yes, it would be better to leave the responsibility of finding the active `generals.csf` file to the Hotkey Editor.

**Comment by xezon** [comment]
Ah ok. Yeah that sounds like it could need some considerations in regards to how the Launcher could pickup user generated content as Addons itself. But I am not sure if that would be even necessary.

**Comment by xezon** [comment]
What we could edit hotkeys for is duplicate assignments. I saw a couple of these in Shockwave change list. I assume every language has individual button conflicts.

**Comment by TranzistorP** [comment]
![изображение](https://user-images.githubusercontent.com/96445511/236879596-26248689-430b-43c8-828c-6a36919918f8.png)

I have an alpha version of the Hotkey Editor. But it can\`t works with .big yet and hasn\`t localizations because I did it only for my mod. But if you need, i can finish it when i have a free time (in July)
